### DISCUSSION TITLE
Trouble with Static Site + Custom domain

### DISCUSSION THREAD

**Discussion by Caio-Domingos** [aws/copilot-cli#5543] [Q&A]
Hi guys! 

## Issue: Static Site + Custom Domain Not Working as Expected with AWS Copilot

**Problem Description:**
I initiated copilot with the command `copilot app init --domain <my domain>`, and my service and environment were created without any issues. When I run the backend service using the Load Balancer Web App, it correctly routes with the URL format `<service>.<env>.<app name>.<domain>`, and the routing works perfectly. However, when I deploy the frontend as a Static Site, it generates two URLs for me: the default Cloudfront URL and the automatically generated domain `<service>.<env>.<app name>.<domain>`. The issue is that only the Cloudfront route is functional.

**Steps Taken:**
- Checked Route 53, where I found the A record pointing as an alias to Cloudfront.
- Attempted to add my own style of alias.
- Used DNS Checker to verify if updates were occurring.
- Left the service active for 2 days to check if DNS propagation was delayed.

Despite these steps, the issue persists, and I need assistance in resolving this.

**Additional Context:**
- The backend service routing works correctly with `<service>.<env>.<app name>.<domain>`.
- Only the Cloudfront URL is operational for the frontend Static Site.

Any guidance or suggestions on how to fix this would be greatly appreciated. Thank you!

![image](https://github.com/aws/copilot-cli/assets/43622756/f3167f46-f1e3-4f3a-a3b6-5ae07785348f)
![image](https://github.com/aws/copilot-cli/assets/43622756/dc41e31a-73fa-41ae-bdd0-00d120c24e59)
![image](https://github.com/aws/copilot-cli/assets/43622756/863686e1-6e86-44a0-b7be-6f89718392be)

![image](https://github.com/aws/copilot-cli/assets/43622756/5be0c8aa-1f88-4647-b469-ecbad5200d1a)

**Comment by iamhopaul123** [comment]
Hello @Caio-Domingos. This looks strange to me 🤔  Can you help me check the route53 record for `frontend.develop.ga-awscop.thinkless.com.br` and see if the A-record is correctly set there (in your app account)?

**Reply by Caio-Domingos** [reply]
Hello, and thank you for your prompt response.

I appreciate your help in troubleshooting this issue. To provide more clarity, I will share screenshots of both the Route 53 settings and the Cloudfront configuration for `frontend.develop.ga-awscop.thinkless.com.br`. This should give a better understanding of how the A-record is set up and any other relevant configurations.

I'll follow up shortly with these screenshots.

Thanks again for your assistance!

![image](https://github.com/aws/copilot-cli/assets/43622756/e42b40e8-b4da-431c-b7c5-5814e5ab3be9)
![image](https://github.com/aws/copilot-cli/assets/43622756/950add32-b3e8-4ce6-baf2-bbfc22b78a20)

**Reply by iamhopaul123** [reply]
Thank you I think your website works now.

```
curl frontend.develop.ga-awscop.thinkless.com.br                                              
<html>
<head><title>301 Moved Permanently</title></head>
<body>
<center><h1>301 Moved Permanently</h1></center>
<hr><center>CloudFront</center>
</body>
</html>
```

**Reply by Caio-Domingos** [reply]
I'm starting to suspect that this might be related to DNS propagation issues, as indicated by one of my recent screenshots.

To provide more clarity, I'll send additional screenshots that not only show the current state from my perspective when trying to access the site but also highlight the DNS propagation status. This may help us identify if the root cause of the issue is indeed related to DNS propagation delays or inconsistencies.

I appreciate your patience and assistance as we work through this issue. I'll forward the screenshots shortly.

Thank you again for your help.


![image](https://github.com/aws/copilot-cli/assets/43622756/a7e16718-b165-4b18-b42b-a40f89612713)
![image](https://github.com/aws/copilot-cli/assets/43622756/aa04c3f2-7905-4c06-bde9-550f904b13c0)
![image](https://github.com/aws/copilot-cli/assets/43622756/ae7c7333-131e-4d6d-8b7a-cbc1b17e775b)

**Comment by Caio-Domingos** [comment]
Hello,

After conducting extensive testing, I have come to some intriguing conclusions that I would like to share for further insights.

I created an application named "ga-awscop" and an environment "develop" to test some deployments. AWS Copilot, behind the scenes, created two hosted zones in Route 53: "ga-awscop.thinkless.com.br" and "develop.ga-awscop.thinkless.com.br". Here's what I found:

1. **Services within "ga-awscop.thinkless.com.br"**: Services like "api.ga-awscop.thinkless.com.br" work flawlessly, with no issues whatsoever.

2. **Services within "develop.ga-awscop.thinkless.com.br"**: However, when I use the "develop.ga-awscop.thinkless.com.br" zone, for example, "frontend.develop.ga-awscop.thinkless.com.br", it fails to function. This also applies if I launch the same "api.ga-awscop.thinkless.com.br" service with the route "api.develop.ga-awscop.thinkless.com.br".

Initially, I thought it might be an issue with the VPC of the environment, but this didn't make sense since a Static Site type service doesn't typically involve VPC interactions, as it's a straightforward S3 -> Cloudfront -> Route 53 (Alias) -> World process. So, I ruled out the VPC issue.

Here are the specific tests I conducted and their outcomes:

- Creating an alias within my hosted zone "thinkless.com.br" => Successful.
- Creating an alias within the hosted zone "ga-awscop.thinkless.com.br" => Successful.
- Services that work within "ga-awscop.thinkless.com.br" do not function when deployed under "develop.ga-awscop.thinkless.com.br" (regardless of the service type, including Static Site/Request-Driven, etc.).
- Manually creating a hosted zone named "caio-testes.thinkless.com.br" => Unsuccessful.

This leads me to believe that these hosted zones, specifically "develop.ga-awscop.thinkless.com.br" and "caio-testes.thinkless.com.br", although created, are non-functional for some reason beyond my current understanding as a developer.

What are your thoughts on this? Any insights or suggestions would be greatly appreciated.

Thank you.

**Reply by Lou1415926** [reply]
Hey @Caio-Domingos Thanks for all the testings! 

> Services that work within "ga-awscop.thinkless.com.br" do not function when deployed under "develop.ga-awscop.thinkless.com.br" (regardless of the service type, including Static Site/Request-Driven, etc.).

**Can you go to _"ga-awscop.thinkless.com.br"'s hosted zone_, and see if there is an NS record for the "develop.ga-awscop.thinkless.com.br" domain? If there is, does the value match with the NS record for "develop.ga-awscop.thinkless.com.br" in the _"develop.ga-awscop.thinkless.com.br" hosted zone_?**

This record is to delegate the management of the domain "develop.ga-awscop.thinkless.com.br" from the hosted zone named "ga-awscop.thinkless.com.br" hosted zone to the hosted zone named "develop.ga-awscop.thinkless.com.br". The A-records that live inside of the "develop.ga-awscop.thinkless.com.br" hosted zone are only useful if this delegation has happened. By default, Copilot does the delegation for you, but I want to double check and rule out the possibility that the delegation somehow didn't happen.

**Reply by Lou1415926** [reply]
I also have a few follow-up questions regarding the experiments.

> Services that work within "ga-awscop.thinkless.com.br" do not function when deployed under "develop.ga-awscop.thinkless.com.br" (regardless of the service type, including Static Site/Request-Driven, etc.).

By "do not function", did you mean that you couldn't `curl frontend.develop.ga-awscop.thinkless.com.br`, or that the propagation only happened partially as in [this comment](https://github.com/aws/copilot-cli/discussions/5543#discussioncomment-7833639) or that the propagation did not happen at all (e.g. after checking the domain on a website like https://dnschecker.org/, the result showed red "X"s all over the world)?

For context, I am trying to understand this experiment together with https://github.com/aws/copilot-cli/discussions/5543#discussioncomment-7833639. The screenshot in [your second command](https://github.com/aws/copilot-cli/discussions/5543#discussioncomment-7833639) indicated that it was propagated partially (which is 100% still an issue, I totally agree, but the solution would be different from if the propagation did not happen at all). 

> Manually creating a hosted zone named "caio-testes.thinkless.com.br" => Unsuccessful.

Can you clarify the steps taken in this experiment? For example, did you \
1) import a certificate for "caio-testes.thinkless.com.br" 
2) add the A record that route "caio-testes.thinkless.com.br" to the destination (perhaps the cloudfront domain, in your case) to the hosted zone? 

For any alias that are not "svc.env.app.domain.com", "env.app.domain.com", "app.domain.com" or "domain.com", Copilot expects you to manage the A records yourself, or use the [hosted_zone](https://aws.github.io/copilot-cli/docs/manifest/lb-web-service/#http-hosted-zone) field in the manifest (the field is not supported for static site though).

**Reply by Lou1415926** [reply]
Finally, just to share - I did a small test myself too: 
1) I created an app with copilot app init --domain [mydomain.com](http://mydomain.com/), and then
2) deploy a Load-Balanced Web Service "web" to the "test" environment. 

I was able to access my service at https://web.test.myapp.mydomain.com/, almost immediately after deployment. A some-what relevant but certainly not decisive factor: the application "myapp" and the "test" environment all live inside of "us-east-1", and I'm access the URL from "us-west-2".

There are two things that are different in my result and yours:
1. The environment-level domain worked for me, but is not working for you.
> Services that work within "[ga-awscop.thinkless.com.br](http://ga-awscop.thinkless.com.br/)" do not function when deployed under "[develop.ga-awscop.thinkless.com.br](http://develop.ga-awscop.thinkless.com.br/)" (regardless of the service type, including Static Site/Request-Driven, etc.).
2. The DNS propagation was quick for me, but was slow for you.
> Left the service active for 2 days to check if DNS propagation was delayed.

----
I don't have a great guess of why you were facing the issue, as I have yet to reproduce the same situation. All of my questions ⬆️ 
 for you were more or less shots in the dark 😞 I'd need more information hoping to find the root cause ASAP 🤞🏼
